Add MAX77659 MFD Driver

ZeynepArslanbenzer opened the pull request:
Add driver support for MAX77659 PMIC.

Signed-off-by: ZeynepArslanbenzer <Zeynep.Arslanbenzer@maximintegrated.com>

nunojsa commented on drivers/mfd/max77659.c:
```diff
@@ -0,0 +1,206 @@
+	return max77659_pmic_setup(me);
+}
+
+static const struct of_device_id max77659_of_id[] = {
```
this requires mod_devicetable.h

nunojsa replied:
mod_devicetable.h still missing... Note we should not rely on other headers to include it

nunojsa commented on drivers/mfd/max77659.c:
```diff
@@ -0,0 +1,206 @@
+	me->regmap = devm_regmap_init_i2c(client, &max77659_regmap_config);
+	if (IS_ERR(me->regmap)) {
+		ret = PTR_ERR(me->regmap);
+		return dev_err_probe(&client->dev, ret, "Failed to allocate register map\n");
```
Maybe just use `PTR_ERR` directly

nunojsa commented on drivers/mfd/max77659.c:
```diff
@@ -0,0 +1,206 @@
+		return dev_err_probe(dev, ret,
+				     "Unable to read Charger Interrupt Status register\n");
+
+	ret = device_init_wakeup(dev, true);
```
Should we call this if `me->irq <= 0` ?

nunojsa commented on drivers/mfd/max77659.c:
```diff
@@ -0,0 +1,206 @@
+
+	return devm_regmap_add_irq_chip(dev, me->regmap, me->irq,
+					IRQF_ONESHOT | IRQF_SHARED, 0,
+					&max77659_chg_irq_chip, &me->irqc_chg);
```
again, I'm not sure if any of this make if `me->irq <= 0`.  I think `devm_regmap_add_irq_chip()` will fail. This to say this code probably does not make much sense:

```
if (me->irq > 0)
		enable_irq_wake(me->irq);
```

I mean, the check is probably meaningless because irq is a required  for the whole thing to work

nunojsa commented on drivers/mfd/max77659.c:
```diff
@@ -0,0 +1,206 @@
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/of_device.h>
+#include <linux/of_irq.h>
```
At fiirst glance, all the the of includes are not really needed... Please double check other ones that are not needed

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+#define MAX77659_BIT_CNFG_C_TOPOFFTIMER GENMASK(2, 0)
+#define MAX77659_BIT_CNFG_E_CC GENMASK(7, 2)
+#define MAX77659_BIT_CNFG_E_TFASTCHG GENMASK(1, 0)
+#define MAX77659_BIT_CNFG_G_CHG_CV GENMASK(7, 2)
```
I would use a tab here. With a single space makes the code a bit hard to read...

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+};
+
+enum{
+	MAX77659_CHG_IRQ_THM_I = 0,
```
no need to set it to 0

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+	unsigned int topoff_timer_min;
+};
+
+static char *chg_supplied_to[] = {
```
`const`?

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+
+	fast_charge_current_ua = clamp_val(fast_charge_current_ua, MAX77659_CHARGER_CURRENT_MIN,
+					   MAX77659_CHARGER_CURRENT_MAX);
+	reg_data = (fast_charge_current_ua / MAX77659_CHARGER_CURRENT_STEP) - 1;
```
no need for parenthesis...

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+	ret = regmap_update_bits(charger->regmap, MAX77659_REG_CNFG_CHG_E,
+				 MAX77659_BIT_CNFG_E_CC, reg_data);
+	if (ret)
+		dev_err(charger->dev, "Error in writing register CNFG_CHG_E\n");
```
I would just `return regmap_update_bits()`

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+		return ret;
+
+	reg_data = (FIELD_GET(MAX77659_BIT_CNFG_E_CC, reg_data) + 1)
+		    * MAX77659_CHARGER_CURRENT_STEP;
```
nit: new line

pcercuei replied:
I'd prefer an extra variable here. E.g.:
```
reg_data = FIELD_GET(MAX77659_BIT_CNFG_E_CC, reg_data);
current = (reg_data + 1) * MAX77659_CHARGER_CURRENT_STEP;
```

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+	unsigned int reg_data;
+
+	termination_time = clamp_val(termination_time, 0, (MAX77659_BIT_CNFG_C_TOPOFFTIMER * 5));
+	reg_data = FIELD_PREP(MAX77659_BIT_CNFG_C_TOPOFFTIMER, (termination_time / 5));
```
no need for parenthesis...

pcercuei replied:
You either need a comment to explain what's the magic value "5" is for, or use a macro for it, with a name that is suggestive enough.

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+
+	ret = of_property_read_u32(np, "topoff-timer", &charger->topoff_timer_min);
+	if (ret)
+		charger->topoff_timer_min = 30;
```
Use device properties... I think all of these can be done with them. dt bindings file is also necessary

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+
+	charger->irq = irq;
+	max77659_charger_update(charger);
+	schedule_delayed_work(&charger->irq_work, MAX77659_IRQ_WORK_DELAY);
```
threaded irq (instead of irq + worker)

pcercuei replied:
Absolutely, you need a threaded IRQ here. And you cannot call `max77659_charger_update` in `max77659_charger_isr`, because `regmap_read` can lock a mutex, and you cannot have that in an IRQ handler (unless it's a threaded IRQ handler).

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+
+		if (charger->irq_arr[i] < 0) {
+			dev_err(&pdev->dev, "Failed to get virq for MAX77659_CHG_IRQ%d\n", i);
+			continue;
```
Does it make sense to just continue if we get an error here? Is that a valid scenario?

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+	{ }
+};
+MODULE_DEVICE_TABLE(platform, max77659_charger_id);
+
```
Add an of_device table...

nunojsa commented on drivers/regulator/max77659-regulator.c:
```diff
@@ -0,0 +1,163 @@
+#include <linux/regulator/driver.h>
+#include <linux/regulator/of_regulator.h>
+#include <linux/slab.h>
+#include <linux/version.h>
```
it looks like we have more headers than needed...

nunojsa commented on drivers/regulator/max77659-regulator.c:
```diff
@@ -0,0 +1,163 @@
+	rdata->initdata = of_get_regulator_init_data(dev, np, &max77659_LDO_desc);
+	rdata->of_node = np;
+
+	of_node_put(np);
```
Can you check if this can be done with device properties? Moreover `of_node_put()` needs to be called if the memory allocation fails...

This driver also needs a dt bindings file

nunojsa replied:
I don't think you need this `max77659_regulator_parse_dt()` at all.... You can set [regulators_node](https://elixir.bootlin.com/linux/latest/C/ident/regulators_node) to the nodename ("regulator" in this case) and together with having `config.dev = pdev->dev.parent` should make sure that all your initdata is initialized during `devm_regulator_register()`

nunojsa commented on drivers/regulator/max77659-regulator.c:
```diff
@@ -0,0 +1,163 @@
+	config.regmap = regmap;
+
+	regulator->rdev = devm_regulator_register(&pdev->dev, &max77659_LDO_desc, &config);
+
```
no need for blank line

nunojsa commented on drivers/regulator/max77659-regulator.c:
```diff
@@ -0,0 +1,163 @@
+	{ }
+};
+MODULE_DEVICE_TABLE(platform, max77659_regulator_id);
+
```
of_device table...

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+	else if (charger->fast_charge_timer_hr <= 7)
+		val = (int)DIV_ROUND_UP(charger->fast_charge_timer_hr - 1, 2);
+	else
+		val = 0x00;
```
I think you could reduce this to 3 cases, the first one being `if (charger->fast_charge_timer_hr == 0 || charger->fast_charge_timer_hr > 7)`

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+static int max77659_charger_parse_dt(struct max77659_charger_dev *charger)
+{
+	struct device *dev = charger->dev;
+	struct device_node *np = of_find_node_by_name(NULL, "charger");
```
Why the NULL?

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+	/* Charger */
+	if (!np)
+		return dev_err_probe(dev, -ENODEV,
+				     "Failed to find device_node\n");
```
The `dev_err_probe` is handy because it doesn't print any message in case of a -EPROBE_DEFER. When you know that this cannot happen, you can just use a regular `dev_err` + return -ENODEV, especially  when you need to break lines anyway.

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+
+	ret = of_property_read_u32(np, "fast-charge-timer", &charger->fast_charge_timer_hr);
+	if (ret)
+		charger->fast_charge_timer_hr = 0;
```
Your `charger->fast_charge_timer_hr` is already zero, so you don't need to set it on error.

By the way, `of_property_read_u32` will return -EINVAL if the property is missing, but you should still handle every other error.

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+
+	ret = of_property_read_u32(np, "fast-charge-current", &charger->fast_charge_current_ua);
+	if (ret)
+		charger->fast_charge_current_ua = 15000;
```
I would suggest to set these defaults when the "charger" is created, and let DT override them instead of setting them here.

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+
+	charger->health = POWER_SUPPLY_HEALTH_UNKNOWN;
+	charger->status = POWER_SUPPLY_STATUS_UNKNOWN;
+	charger->charge_type = POWER_SUPPLY_CHARGE_TYPE_UNKNOWN;
```
These fields are set in every case except when the regmap_read fails, which shouldn't really happen. So I'd suggest to move these three lines in the error handling case of the regmap_read.

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+		ret = max77659_get_charge_current(charger, &val->intval);
+		break;
+	default:
+		ret = -EINVAL;
```
break?

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+		charger->fast_charge_current_ua = val->intval;
+		break;
+	default:
+		ret = -EINVAL;
```
break here too.

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+					   MAX77659_BIT_CNFG_B_CHG_EN, true);
+		} else {
+			regmap_update_bits(charger->regmap, MAX77659_REG_CNFG_CHG_B,
+					   MAX77659_BIT_CNFG_B_CHG_EN, false);
```
I'd prefer to have:
```
regmap_update_bits(charger->regmap, MAX77659_REG_CNFG_CHG_B,
                   MAX77659_BIT_CNFG_B_CHG_EN, !!chg_present);
```

Then you only have one call  to regmap_update_bits and it reduces a bit the complexity of the code.

If you choose to keep the current code, then `chg_present` should be a `bool` since it's used as such.

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+	{ "charger-chgin" },
+	{ "charger-tj-reg" },
+	{ "charger-sys-ctrl" },
+};
```
First of all, your array is non-const and it doesn't have a valid reason to be mutable.

Second... Why not just an array of strings?
`static const char * const max77659_irq_descs[] = { ... };`

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+
+	charger = devm_kzalloc(&pdev->dev, sizeof(*charger), GFP_KERNEL);
+	if (!charger)
+		return dev_err_probe(&pdev->dev, -ENOMEM, "Memory allocation failed\n");
```
Same as before - no error message on allocation failure. Just return -ENOMEM here.

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+
+	ret = max77659_charger_parse_dt(charger);
+	if (ret)
+		return dev_err_probe(&pdev->dev, ret, "Not found charger dt!\n");
```
"Not found charger dt" isn't very English ;)
And `max77659_charger_parse_dt` already prints a message on error, so you can just `return ret;` here.

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+	int i, ret;
+	struct power_supply_config charger_cfg = {};
+
+	charger = devm_kzalloc(&pdev->dev, sizeof(*charger), GFP_KERNEL);
```
Keep a `struct device *dev` variable in your probe and use it instead of `&pdev->dev`, since I see this pattern everywhere in this function.

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+						max77659_irq_descs[i].name, charger);
+		if (ret)
+			return dev_err_probe(&pdev->dev, ret, "Failed to request irq: %d\n",
+									charger->irq_arr[i]);
```
Use braces if the content of the conditional is multi-lines, which is the case here.

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+						      &charger->psy_chg_d, &charger_cfg);
+	if (IS_ERR(charger->psy_chg))
+		return dev_err_probe(&pdev->dev, PTR_ERR(charger->psy_chg),
+							 "Failed to register power supply\n");
```
Same here.

pcercuei commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,585 @@
+
+static const struct platform_device_id max77659_charger_id[] = {
+	{ MAX77659_CHARGER_NAME, 0, },
+	{ }
```
Add `/* sentinel */` inside the empty braces. I personally don't care but some maintainers do want this.

pcercuei commented on drivers/regulator/max77659-regulator.c:
```diff
@@ -0,0 +1,163 @@
+static int max77659_regulator_parse_dt(struct max77659_regulator_dev *regulator)
+{
+	struct device *dev = regulator->dev;
+	struct device_node *np = of_find_node_by_name(NULL, "regulator");
```
Why the NULL?

pcercuei commented on drivers/regulator/max77659-regulator.c:
```diff
@@ -0,0 +1,163 @@
+
+	rdata = devm_kzalloc(dev, sizeof(*rdata), GFP_KERNEL);
+	if (!rdata)
+		return dev_err_probe(dev, -ENOMEM, "Failed to allocate memory\n");
```
Just return -ENOMEM

pcercuei commented on drivers/regulator/max77659-regulator.c:
```diff
@@ -0,0 +1,163 @@
+
+	regulator = devm_kzalloc(&pdev->dev, sizeof(struct max77659_regulator_dev), GFP_KERNEL);
+	if (!regulator)
+		return dev_err_probe(&pdev->dev, -ENOMEM, "Failed to allocate memory\n");
```
`sizeof(*regulator)` and just return -ENOMEM

pcercuei commented on drivers/regulator/max77659-regulator.c:
```diff
@@ -0,0 +1,163 @@
+	regulator->rdev = devm_kzalloc(&pdev->dev, sizeof(struct regulator_dev *), GFP_KERNEL);
+	if (!regulator->rdev)
+		return dev_err_probe(&pdev->dev, -ENOMEM,
+		"Failed to allocate memory for regulator devices\n");
```
Same as above.

pcercuei commented on drivers/regulator/max77659-regulator.c:
```diff
@@ -0,0 +1,163 @@
+
+	if (IS_ERR(regulator->rdev))
+		return dev_err_probe(&pdev->dev, PTR_ERR(regulator->rdev),
+				     "Failed to register regulator %s\n", max77659_LDO_desc.name);
```
Need braces.

pcercuei commented on drivers/regulator/max77659-regulator.c:
```diff
@@ -0,0 +1,163 @@
+static struct platform_driver max77659_regulator_driver = {
+	.driver = {
+		.name = MAX77659_REGULATOR_NAME,
+	},
```
You can write `.driver.name = MAX77659_REGULATOR_NAME,` unless you expect more fields to be added afterwards.
Not a big deal anyway.

pcercuei commented on include/linux/mfd/max77659.h:
```diff
@@ -0,0 +1,59 @@
+    struct regmap *regmap;
+};
+
+#endif /* __MAX77659_MFD_H__ */
```
Github says "No newline at end of file", looks like your last line does not have a carriage return (I don't mean an empty line).

pcercuei commented on include/linux/mfd/max77659.h:
```diff
@@ -0,0 +1,59 @@
+    struct regmap_irq_chip_data	*irqc_chg;
+
+    struct i2c_client *i2c;
+    struct regmap *regmap;
```
This header should not depend on all these structures being already defined. Thanksfully they are only used as opaque pointers, so before your `struct max77659_dev` definition, you can add:
```
struct device;
struct i2c_client;
struct regmap_irq_chip_data;
struct regmap;
```

Then you won't need any includes in this file.

pcercuei commented on include/linux/mfd/max77659.h:
```diff
@@ -0,0 +1,59 @@
+#define MAX77659_INT_GLBL0_MASK 0xFF
+#define MAX77659_INT_GLBL1_MASK 0x33
+
+#define MAX77659_BIT_INT_GLBL0_GPIO0_F BIT(0)
```
You are using the `BIT()` macro so you need to include `<linux/bits.h>` to use the macro.

pcercuei commented on include/linux/mfd/max77659.h:
```diff
@@ -0,0 +1,59 @@
+#define MAX77659_BIT_INT_SYS_CTRL BIT(4)
+
+enum {
+    MAX77659_DEV_PMIC = 0,
```
No need for the `= 0`, this is the default.

nunojsa commented:
> On top of this, for a MFD device, you should have one commit per driver (naturally with the MFD driver being the first one) instead of one big commit adding all drivers.

As stated, please add one commit per driver updating the MAINTAINERS file along the way...

> add proper bindings for the drivers.

devicetree bindings are also needed 

Also address valid checkpatch warnings. From a quick look I could already see things like:

```
if (ret) {
      dev_err_probe()
}
```

This will trigger a warning since the sytle is not to have {} on one line statements... On top of this, you also have a mixed style regarding this (some places you have the braces, some you don't)

ZeynepArslanbenzer commented:
> > On top of this, for a MFD device, you should have one commit per driver (naturally with the MFD driver being the first one) instead of one big commit adding all drivers.
> 
> As stated, please add one commit per driver updating the MAINTAINERS file along the way...
> 
> > add proper bindings for the drivers.
> 
> devicetree bindings are also needed
> 
> Also address valid checkpatch warnings. From a quick look I could already see things like:
> 
> ```
> if (ret) {
>       dev_err_probe()
> }
> ```
> 
> This will trigger a warning since the sytle is not to have {} on one line statements... On top of this, you also have a mixed style regarding this (some places you have the braces, some you don't)

ZeynepArslanbenzer commented:
* I have added one commit for each driver.
* I have added device tree bindings.
* I have updated the MAINTAINERS file.
* @nunojsa I added braces if the content of the conditional is multi-lines, as mentioned in this conversation -> https://github.com/analogdevicesinc/linux/pull/1982#discussion_r964731700

nunojsa commented on drivers/mfd/max77659.c:
```diff
@@ -0,0 +1,200 @@
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to add sub-devices\n");
+
+	if (me->irq > 0)
```
don't need this check right? If we do not have any irq, `devm_regmap_add_irq_chip()` should fail

nunojsa commented on drivers/mfd/max77659.c:
```diff
@@ -0,0 +1,200 @@
+
+static const struct i2c_device_id max77659_i2c_id[] = {
+	{ MAX77659_NAME, 0 },
+	{ /* sentinel */ }
```
nit: I would remove the `/* sentinel */` comment...

pcercuei replied:
And I would keep it ;) Some maintainers prefer to have it.

nunojsa replied:
well, it is a matter of preference yes :)

nunojsa commented on drivers/mfd/max77659.c:
```diff
@@ -0,0 +1,200 @@
+	int ret;
+
+	if (client->irq <= 0)
+		return dev_err_probe(dev, -EINVAL, "Invalid IRQ\n");
```
no need for this check... just let things fail normally (at `devm_regmap_add_irq_chip()`)

nunojsa commented on drivers/regulator/max77659-regulator.c:
```diff
@@ -0,0 +1,150 @@
+	config.driver_data = regulator;
+	config.init_data = regulator->rdata->initdata;
+	config.of_node = regulator->rdata->of_node;
+	config.regmap = regmap;
```
newline

nunojsa commented on drivers/regulator/max77659-regulator.c:
```diff
@@ -0,0 +1,150 @@
+	config.dev = &pdev->dev;
+	config.driver_data = regulator;
+	config.init_data = regulator->rdata->initdata;
+	config.of_node = regulator->rdata->of_node;
```
don't think you need this one if you try my approach...

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,551 @@
+	else if (charger->fast_charge_timer_hr < 3)
+		val = 0x01;
+	else
+		val = (int)DIV_ROUND_UP(charger->fast_charge_timer_hr - 1, 2);
```
drop the cast

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,551 @@
+{
+	struct max77659_charger_dev *charger = data;
+
+	charger->irq = irq;
```
why the above? looks odd

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,551 @@
+
+	max77659_charger_parse_dt(charger);
+
+	platform_set_drvdata(pdev, charger);
```
not seeing `platform_get_drvdata()`. Likely means you can drop this

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,551 @@
+		return dev_err_probe(dev, ret, "Failed to initialize charger\n");
+
+	for (i = 0; i < MAX77659_CHG_IRQ_MAX; i++) {
+		charger->irq_arr[i] = regmap_irq_get_virq(max77659->irqc_chg, i);
```
Hmm, I think you don't need this... Look at my comment on the main MFD driver

nunojsa commented on drivers/power/supply/max77659-charger.c:
```diff
@@ -0,0 +1,551 @@
+	}
+
+	charger->psy_chg = devm_power_supply_register(dev,
+						      &charger->psy_chg_d, &charger_cfg);
```
if it does not break the 100 col limit I would have a one liner... Not that neat like this

nunojsa commented on drivers/mfd/max77659.c:
```diff
@@ -0,0 +1,200 @@
+		return dev_err_probe(dev, ret, "Unable to init wakeup\n");
+
+	ret = devm_mfd_add_devices(dev, -1, max77659_devs, ARRAY_SIZE(max77659_devs),
+				   NULL, 0, NULL);
```
AFAIC, you can pass in your regmap domain in here with [regmap_irq_get_domain()](https://elixir.bootlin.com/linux/latest/C/ident/regmap_irq_get_domain) with `me->irqc_chg`. Then you don't need to create the mappings (calling `regmap_irq_get_virq()`) in the charger driver.

That said, I see we have 3 irqchips in here so favoring the charger one might be questionable. OTOH, seems to be the only one used... Or maybe this just means that we can have just one irqchip?

nunojsa commented on Documentation/devicetree/bindings/mfd/max77659.yaml:
```diff
@@ -0,0 +1,73 @@
+      The first cell is the IRQ number, the second cell is the trigger type.
+
+  charger:
+    $ref: ../power/supply/max77659-charger.yaml
```
I think we need a similar property for `regulator` since you want to make use of the initdata. Just reference the default regulator.yaml file

nunojsa commented on Documentation/devicetree/bindings/power/supply/max77659-charger.yaml:
```diff
@@ -0,0 +1,40 @@
+  compatible:
+    const: adi,max77659-charger
+
+  fast-charge-timer:
```
This needs a vendor prefix `adi,fast-charge-timer`... Moreover it should come in seconds as this a standard unit. Hence `adi,fast-charge-sec`.

Look at https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/property-units.yaml

nunojsa commented on Documentation/devicetree/bindings/power/supply/max77659-charger.yaml:
```diff
@@ -0,0 +1,40 @@
+
+  fast-charge-current:
+    description:
+      Fast-charge constant current value (in microamps).
```
name should be `adi,fast-charge-microamp`

nunojsa commented on Documentation/devicetree/bindings/power/supply/max77659-charger.yaml:
```diff
@@ -0,0 +1,40 @@
+    description:
+      Fast-charge constant current value (in microamps).
+
+  topoff-timer:
```
`adi,topoff-sec`

nunojsa reviewed (commented):
Alright, a few more comments from me. I would say now to send the patches upstream :)

ZeynepArslanbenzer commented:
We merged the max77659 driver to [this driver](https://lore.kernel.org/lkml/?q=max77658). We sent it to mainline Linux. I will close this pull request. Thank you for your guidance.

nunojsa commented:
> We merged the max77659 driver to this driver. We sent it to mainline Linux. I will close this pull request. Thank you for your guidance.

Closing based on the above...
